addons/web_editor/static/src/js/editor/snippets.editor.js

```diff
@@ -267,61 +318,75 @@ var SnippetEditor = Widget.extend({
+                    $snippet: this.$snippetBlock,
+                    callback: function (editor, $snippet) {
+                        for (var i in editor.snippetOptionInstances) {
+                            editor.snippetOptionInstances[i].onRemove();
+                        }
+                        resolve();
+                    },
+                });
             });
-        });
 
-        this.trigger_up('go_to_parent', {$snippet: this.$target});
-        var $parent = this.$target.parent();
-        this.$target.find('*').addBack().tooltip('dispose');
-        this.$target.remove();
-        this.$el.remove();
-
-        var node = $parent[0];
-        if (node && node.firstChild) {
-            if (!node.firstChild.tagName && node.firstChild.textContent === ' ') {
-                node.removeChild(node.firstChild);
+            this.trigger_up('go_to_parent', {$snippet: this.$snippetBlock});
+            var $parent = this.$snippetBlock.parent();
+            this.$snippetBlock.find('*').addBack().tooltip('dispose');
+            // The snippet has to be removed directly in jquery as well for the
+            // synchronous bubbling up system to remove the empty parent.
+            // Ideally, we should simply call remove inside a mutations observer
+            // callback such that the removal is synchronous while the vdom is
+            // updated through observed mutations without calling remove twice.
+            this.$snippetBlock.remove();
+            await this.editorHelpers.remove(context, this.$snippetBlock[0]);
+            this.$el.remove();
+
+            var node = $parent[0];
+            if (node && node.firstChild) {
+                if (!node.firstChild.tagName && node.firstChild.textContent === ' ') {
+                    await this.editorHelpers.remove(context, node.firstChild);
+                }
             }
-        }
 
-        if ($parent.closest(':data("snippet-editor")').length) {
-            var editor = $parent.data('snippet-editor');
-            while (!editor) {
-                var $nextParent = $parent.parent();
-                if (isEmptyAndRemovable($parent)) {
-                    $parent.remove();
+            if ($parent.closest(':data("snippet-editor")').length) {
+                var editor = $parent.data('snippet-editor');
+                while (!editor) {
+                    var $nextParent = $parent.parent();
+                    if (isEmptyAndRemovable($parent)) {
+                        await this.editorHelpers.remove(context, this.$parent[0]);
+                    }
+                    $parent = $nextParent;
+                    editor = $parent.data('snippet-editor');
+                }
+                if (isEmptyAndRemovable($parent, editor)) {
+                    // TODO maybe this should be part of the actual Promise being
+                    // returned by the function ?
+                    await new Promise((resolve)=> {
```

Comment:
It looks like whoever wrote that meant that we should _return_ this promise. I'm guessing something like pushing this one to an array of promises and then returning `Promises.all(promises)`.

Reply:
No, I probably wrote it. All promises here do not return anything, this was just not awaited before, it is now (because probably possible now?), so the comment should not make sense anymore.

Reply:
Ok I guess you're right then, you would know. If none of the promises return anything it does seem weird indeed to return them... :shrug:

Reply:
I confirm that this comment can be removed. It made sense when @qsm-odoo wrote it but it does not make sense anymore in the new code of @Goaman since we indeed properly wait for it now.